plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +241,185 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            paramsTobeUpdated.put("name", "Nuage VSD - " + command.getHostName());
+            paramsTobeUpdated.put("hostname", command.getHostName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getUserName()) &&
+                !command.getUserName().equals(nuageVspHost.getDetails().get("cmsuser"))) {
+            paramsTobeUpdated.put("cmsuser", command.getUserName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getPassword())) {
+            String encodedNewPassword = org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes()));
+            if (!encodedNewPassword.equals(nuageVspHost.getDetails().get("cmsuserpass"))) {
+                paramsTobeUpdated.put("cmsuserpass", org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes())));
+            }
+        }
+
+        if (command.getPort() != null &&
+                command.getPort() != Integer.parseInt(nuageVspHost.getDetails().get("port"))) {
+            paramsTobeUpdated.put("port", String.valueOf(command.getPort()));
+        }
+
+        GetClientDefaultsCommand getClientDefaultsCmd = new GetClientDefaultsCommand();
+        GetClientDefaultsAnswer getClientDefaultsAnswer = (GetClientDefaultsAnswer) _agentMgr.easySend(nuageVspHost.getId(), getClientDefaultsCmd);
+        String apiVersion = MoreObjects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        Answer supportedApiVersionAnswer = _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
+
+        final int retryCount = Integer.parseInt(nuageVspHost.getDetails().get("retrycount"));
+        if (command.getApiRetryCount() != null && command.getApiRetryCount() != retryCount) {
+            paramsTobeUpdated.put("retrycount", String.valueOf(command.getApiRetryCount()));
+        }
+
+        final int apiRetryInterval = Integer.parseInt(nuageVspHost.getDetails().get("retryinterval"));
+        if (command.getApiRetryInterval() != null && command.getApiRetryInterval() != apiRetryInterval) {
+            paramsTobeUpdated.put("retryinterval", String.valueOf(command.getApiRetryInterval()));
+        }
+
+        if (paramsTobeUpdated.isEmpty()) {
+            if (s_logger.isDebugEnabled()) {
```

Comment:
There's a <code>return</code> statement right underneath the 2nd if block, we can't make the <code>return</code> statement conditional based on the logging level.